expression/builtin_grouping.go

```diff
@@ -0,0 +1,206 @@
+}
+
+func (b *builtinGroupingImplSig) getMetaGroupingID() int64 {
+	var metaGroupingID int64
+	grouping_ids := b.getMetaGroupingMarks()
+	for key := range grouping_ids {
+		metaGroupingID = key
+	}
+	return metaGroupingID
+}
+
+func (b *builtinGroupingImplSig) checkMetadata() error {
+	mode := b.getGroupingMode()
+	grouping_ids := b.getMetaGroupingMarks()
+	if mode != tipb.GroupingMode_ModeBitAnd && mode != tipb.GroupingMode_ModeNumericCmp && mode != tipb.GroupingMode_ModeNumericSet {
+		return errors.Errorf("Mode of meta data in grouping function is invalid. input mode: %d", mode)
+	} else if (mode == 1 || mode == 2) && len(grouping_ids) != 1 {
+		return errors.Errorf("Invalid number of groupingID. mode: %d, number of groupingID: %d", mode, len(b.groupingMarks))
+	}
+	return nil
+}
+
+func (b *builtinGroupingImplSig) groupingImplBitAnd(groupingID int64, metaGroupingID int64) int64 {
+	if groupingID&metaGroupingID > 0 {
+		return 1
+	}
+	return 0
+}
+
+func (b *builtinGroupingImplSig) groupingImplNumericCmp(groupingID int64, metaGroupingID int64) int64 {
+	if groupingID > metaGroupingID {
+		return 1
+	}
+	return 0
+}
+
+func (b *builtinGroupingImplSig) groupingImplNumericSet(groupingID int64) int64 {
+	grouping_ids := b.getMetaGroupingMarks()
+	_, ok := grouping_ids[groupingID]
+	if ok {
+		return 0
+	}
+	return 1
+}
+
+func (b *builtinGroupingImplSig) grouping(groupingID int64) int64 {
+	switch b.mode {
+	case tipb.GroupingMode_ModeBitAnd:
+		return b.groupingImplBitAnd(groupingID, b.getMetaGroupingID())
+	case tipb.GroupingMode_ModeNumericCmp:
+		return b.groupingImplNumericCmp(groupingID, b.getMetaGroupingID())
+	case tipb.GroupingMode_ModeNumericSet:
+		return b.groupingImplNumericSet(groupingID)
+	}
+	return 0
+}
+
+// evalInt evals a builtinGroupingSig.
+func (b *builtinGroupingImplSig) evalInt(row chunk.Row) (int64, bool, error) {
+	err := b.checkMetadata()
+	if err != nil {
+		return 0, false, err
+	}
+
+	groupingID, isNull, err := b.args[0].EvalInt(b.ctx, row)
+	if isNull || err != nil {
+		return 0, isNull, err
+	}
+
+	return b.grouping(groupingID), false, nil
+}
+
+func (b *builtinGroupingImplSig) groupingVec(groupingIds *chunk.Column, rowNum int, result *chunk.Column) {
+	result.ResizeInt64(rowNum, false)
+	resContainer := result.Int64s()
+	switch b.mode {
+	case tipb.GroupingMode_ModeBitAnd:
+		metaGroupingID := b.getMetaGroupingID()
+		for i := 0; i < rowNum; i++ {
+			resContainer[i] = b.groupingImplBitAnd(groupingIds.GetInt64(i), metaGroupingID)
+		}
+	case tipb.GroupingMode_ModeNumericCmp:
+		metaGroupingID := b.getMetaGroupingID()
+		for i := 0; i < rowNum; i++ {
+			resContainer[i] = b.groupingImplNumericCmp(groupingIds.GetInt64(i), metaGroupingID)
+		}
+	case tipb.GroupingMode_ModeNumericSet:
+		for i := 0; i < rowNum; i++ {
+			resContainer[i] = b.groupingImplNumericSet(groupingIds.GetInt64(i))
+		}
+	}
+}
+
+func (b *builtinGroupingImplSig) vecEvalInt(input *chunk.Chunk, result *chunk.Column) error {
```

Comment:
The same as `evalInt`, why not check metadata here?

Reply:
> The same as `evalInt`, why not check metadata here?

I will check it.